x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/FileStructureFinderManager.java

```diff
@@ -283,6 +308,7 @@ FileStructureFinder makeBestStructureFinder(List<String> explanation, String sam
         String line;
         while ((line = bufferedReader.readLine()) != null && ++lineCount <= maxLines) {
             sample.append(line).append('\n');
+            timeoutChecker.check("sample line splitting");
```

Comment:
This is a thin loop to be doing the check every time. There is a similar problem in `DelimitedFileStructureFinder` I don't want to get into premature optimisations but it would be nice if the check wasn't run every iteration. Possibly the overhead of the logic to check that would be heavier than the fast call to `timeoutChecker.check(...)`

Reply:
> Possibly the overhead of the logic to check that would be heavier

That's my assumption - see 
https://github.com/elastic/elasticsearch/pull/34117#discussion_r221234591

I'll profile it to confirm.